### PR TITLE
[Diffusion] MAGI-2: regional compile coverage

### PR DESCRIPTION
**Pull request by cr-gao**
## Purpose

Part of the MAGI-2 Preview roadmap in #7085 (M4: regional compile coverage). The runner already calls `setup_compile()` for every non-eager deployment, and the native MAGI-2 transformer had no `_repeated_blocks`, so compile was silently a no-op for MAGI-2. This PR gives the transformer a regional torch.compile path.

Changes:

- The transformer layer forward is split into region methods around the two eager kernels (packed FlashAttention with Ulysses exchange, and the multi-head MoE routing/expert kernel): `_attention_input`, then `_dense_output` for dense layers or `_moe_input` / `_moe_output` for MoE layers. `Magi2PreviewTransformer.compile_regions()` compiles those methods plus the pre/post adapters with `fullgraph=True`; the pipeline exposes `setup_compile()`, honors `diffusion_compile_dynamic`, and passes `options={"emulate_precision_casts": True}` so inductor keeps eager's intermediate bf16 roundings in the mHC connections. MoE routing, the expert kernel, EP/Ulysses communication and CFG/DLO/Cache-DiT semantics are unchanged. Compile stays default-on; there is no model-specific switch.
- Two eager hygiene commits, both bitwise equal to the merge base: sampler lengths stay host integers instead of a device round trip with `.item()`, and the pre adapter embeds straight into the checkpoint dtype instead of through an fp32 `[tokens, 4 x hidden]` buffer. Reusing the packing metadata (modality dispatcher, index tensors, RoPE) across denoising steps was also measured and not adopted: the denoise step stayed at 1.931 vs 1.932 s and peak memory rose by 0.23 GiB.

## Numerics

The compiled path is deterministic (bitwise reproducible run to run) but not bitwise equal to eager. What remains after the precision-cast emulation is one fp32 ulp in Triton's `exp` / `sigmoid` versus eager's `expf`, which the following bf16 cast turns into a one-ulp flip in about 1e-5 of the elements per op. No inductor option removes it.

MAGI-2 amplifies any such difference: it routes each (head, token) slot to the top 6 of 256 experts, and at the first MoE layer 16% of the slots have a score margin below 1e-3 between the 6th and 7th expert. On A800 (resident SP4, 272p, seed 42, T2VA), the compiled path changes the expert set of 3.7% of the slots at the first MoE layer and 43% over all 36 MoE layers at the first step. Two eager runs with `MAGI2_DETERMINISTIC=0` (kernel ordering only) differ by 42%, and a deterministic eager run with one bf16 ulp added to one element of the initial noise differs by 43%. The compiled-vs-eager difference is of the same magnitude as kernel reduction order.

Consequently the 100-step T2VA video metric does not resolve implementation differences: three same-code eager pairs land at 23.9, 25.5 and 27.8 dB PSNR, the one-ulp noise change at 28.3 dB, and compiled vs eager at 24.3 dB. Frames 0 / 62 / 124, rows: deterministic eager, eager with the one-ulp noise change, compiled, two non-deterministic eager runs. Same scene, same fox, same motion, different walk phase per row, no artifacts:

T2VA frames

I2VA (reference image, speaking prompt), rows: deterministic eager, compiled:

I2VA frames

## Acceptance criteria and status

A800 resident SP4, 272p, seed 42, `MAGI2_DETERMINISTIC=1`, eager vs compiled:

| criterion | measured | status |
|---|---:|---|
| one step, T2VA: PSNR not below the one-step eager self-difference with `MAGI2_DETERMINISTIC=0` (34.3 dB) | 35.3 dB | met |
| 100 steps, audio cosine >= 0.99, T2VA and I2VA | 0.998, 0.994 | met |
| 100 steps, I2VA: SSIM >= 0.95 and PSNR >= 30 dB | 0.963, 35.7 dB | met |
| 100 steps, T2VA video: no threshold, frames above | 24.3 dB (same-code pairs 23.9 to 27.8 dB) | reported |

## Compatibility

| feature | compile path |
|---|---|
| Ulysses SP (SP4, TP1) | tested on A800, all numbers above |
| DLO rank-local SP2, DLO AllGather DP2 x SP2, HSDP4 | tested on the tiny config (4-rank gloo): compiled bitwise equal to uncompiled in the same mode |
| TP > 1, CFG parallel, layerwise CPU offload, Cache-DiT, quantization | untested with compile |

## Speed and memory

A800 (4x SXM4 80 GB), resident SP4, 272p, 20 steps, `MAGI2_DETERMINISTIC=1`, one warmup request then 3 measured requests per session, base run twice for the noise floor. Median over the measured requests:

| metric | base run 1 | base run 2 | PR eager | PR compiled |
|---|---:|---:|---:|---:|
| denoise step (s) | 1.929 | 1.935 | 1.931 | 1.481 |
| full response (s) | 82.4 | 82.0 | 86.8 | 76.4 |
| peak reserved per rank, max (GiB) | 65.53 | 65.53 | 65.76 | 64.41 |
| first (warmup) request (s) | 100.0 | 95.1 | 103.9 | 138.5 |

The compiled denoise step is 23.3% below base, 75x the base-vs-base gap; per-rank peak memory is 1.1 to 2.2 GiB lower on every rank; the first request pays about 38 s of inductor compilation. The eager commits show no measurable difference (the full-response spread sits in the transformer-to-host staging noise of 34 to 45 s per request).

## Test Plan

**vLLM Version:** 0.28.0

```
PYTHONPATH=$PWD python -m pytest tests/diffusion/models/magi2/ -q
```

## Test Result

- 55 passed: CPU suites, three 4-rank gloo tests (compile under DLO/HSDP, TP/SP parity, DLO all-gather) and the CUDA compile test (RTX 4070 Laptop, torch 2.13.0+cu130)
- Compile tests: regions fullgraph with no graph break, bitwise equal to eager on the tiny config, 7 unique graphs, no recompile across steps, a second request and a longer text under `error_on_recompile`
- A800 real weights: one-step eager output bitwise equal to the merge base; 100-step deterministic eager bitwise reproducible
- ruff, mypy and CI pass

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
This PR touches **tests/diffusion/, vllm_omni/diffusion/** (8 files). Based on CODEOWNERS coverage of the changed files, the most-related reviewers appear to be:

> @Bounty-hunter @fhfuih @wtomin

Could one of you take a look when you get a chance? Thanks!

**Comment by cr-gao** [comment]
Self-review, what I checked before marking ready:

- **Stack**: rebased on current `main` (38abc0ba2), 4 commits, no longer stacked on #7112 (closed; its metadata reuse is dropped, its two hygiene commits are the first two here). The compiled regions were verified not to depend on the reuse: fresh dispatcher, index tensors and RoPE every step still give 7 graphs, no recompile, no graph break on CPU, RTX 4070 inductor and 4-rank gloo.
- **Eager parity on real weights**: one-step eager output sha256 equal to the merge base; 100-step deterministic eager T2VA and I2VA bitwise reproducible and identical to the runs at the previous head. The removed block-entry cast was a no-op after the pre adapter change.
- **Compiled path**: deterministic (routes and outputs bitwise across reruns and across both heads), not bitwise equal to eager. The difference is at the ulp level: first-step MoE route flips 43.4% vs 41.6% for two eager runs with non-deterministic kernels and 43.1% for one bf16 ulp in one noise element; one-step PSNR 35.32 dB vs 34.31 dB eager self-difference; 100-step audio cosine 0.998 / 0.994 (T2VA / I2VA); 100-step I2VA 35.74 dB / SSIM 0.963.
- **T2VA 100-step video**: no threshold claimed. Three same-code eager pairs land at 23.9 / 25.5 / 27.8 dB and a one-ulp input change at 28.3 dB, so that metric does not resolve ulp-level differences; frames are attached instead.
- **Speed and memory** (A800 SP4, 272p, 20 steps, head 385efa63c): denoise step 1.481 vs 1.932 s (23.3% less, 75x the base-vs-base gap), per-rank peak 1.1 to 2.2 GiB lower, first request +38 s compile. The eager commits change nothing measurable.
- **Compatibility**: SP4 on A800; DLO SP2, DLO AllGather DP2xSP2 and HSDP4 on the tiny config (compiled bitwise equal to uncompiled). TP > 1, CFG parallel, layerwise offload, Cache-DiT and quantization are untested with compile and marked so in the table.
- **Tests**: whole `tests/diffusion/models/magi2/` directory 55 passed locally (CPU, 4-rank gloo, CUDA); pre-commit and CI green.

**Comment by cr-gao** [comment]
Self-review (ai assisted), what I checked:

- **Eager parity on real weights**: the two eager commits leave the one-step output bitwise equal to the merge base; 100-step deterministic eager is bitwise reproducible.
- **Compiled path**: deterministic run to run, not bitwise equal to eager. The difference is at the ulp level and of the same magnitude as kernel reduction order: first-step MoE route flips 43% (compiled) vs 42% (two eager runs with non-deterministic kernels) vs 43% (one bf16 ulp in one noise element); one-step PSNR 35.3 dB vs 34.3 dB eager self-difference; 100-step audio cosine 0.998 / 0.994 (T2VA / I2VA); 100-step I2VA 35.7 dB / SSIM 0.963.
- **T2VA 100-step video**: no threshold claimed, because three same-code eager pairs land at 23.9 to 27.8 dB and a one-ulp input change at 28.3 dB; frames are attached instead.
- **Speed and memory** (A800 SP4, 272p, 20 steps): denoise step 1.481 vs 1.932 s, per-rank peak 1.1 to 2.2 GiB lower, first request +38 s compile. The eager commits change nothing measurable.
- **Compatibility**: SP4 on A800; DLO SP2, DLO AllGather DP2xSP2 and HSDP4 on the tiny config. TP > 1, CFG parallel, layerwise offload, Cache-DiT and quantization are untested with compile and marked so.
- **Tests**: whole `tests/diffusion/models/magi2/` directory 55 passed locally (CPU, 4-rank gloo, CUDA); pre-commit and CI green.

File: vllm_omni/diffusion/models/magi2/pipeline_magi2.py [hunk 1]
```diff
@@ -695,6 +695,17 @@ def vae(self) -> nn.Module:
 
         return self.video_decoder.module
 
+    def setup_compile(self) -> None:
+        """Compile the transformer regions; the attention and MoE kernels stay eager."""
```
Comment (yeahdongcn) [inline vllm_omni/diffusion/models/magi2/pipeline_magi2.py:699]: Should we also provide a way to configure the Dynamo recompile limit? I encountered an error on a Moore Threads GPU.

Ref: https://github.com/kissablemt/dinov3/blob/main/fsdp_to_single.py#L33-L35

Comment (cr-gao) [reply vllm_omni/diffusion/models/magi2/pipeline_magi2.py:699]: Could you share the torch version or some detailed traceback? If it is because of the torch version I think we can add a config knob in the common diffusion compile config, with a separate follow-up issue & PR.

Comment (yeahdongcn) [reply vllm_omni/diffusion/models/magi2/pipeline_magi2.py:699]: I believe this value has been set to `8`: https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/config.py#L121 for years. We could make it configurable in a separate PR, or I can include that change as part of the Moore Threads GPU support.

Comment (yeahdongcn) [reply vllm_omni/diffusion/models/magi2/pipeline_magi2.py:699]: https://github.com/vllm-project/vllm-omni/pull/7243

File: vllm_omni/diffusion/models/magi2/pipeline_magi2.py [hunk 2]
```diff
@@ -695,6 +695,17 @@ def vae(self) -> nn.Module:
 
         return self.video_decoder.module
 
+    def setup_compile(self) -> None:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/magi2/pipeline_magi2.py:698]: `setup_compile()` forwards `diffusion_compile_dynamic` but silently ignores `diffusion_compile_granularity`, so an explicit `--diffusion-compile-granularity full` request quietly compiles regional instead. `regional_compilation.md` does say `setup_compile()` pipelines manage their own policy — a warn (or assert) on `granularity != "regional"` here would make the downgrade visible.

File: tests/diffusion/models/magi2/test_native_compile_cuda.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    with torch._dynamo.config.patch(error_on_recompile=True):
+        compiled = [first, sampler.forward(steps[1]), sampler.forward(steps[2])]
+        repeated = [sampler.forward(step) for step in steps]
+        # A second request repacks the same shapes; a third changes the
+        # text length and therefore every token count.
+        compiled_new_request = sampler.forward(new_request_step)
+        compiled_longer_text = sampler.forward(longer_text_step)
+
+    for (video, audio), (video_ref, audio_ref) in zip(compiled, expected, strict=True):
+        assert torch.equal(video, video_ref)
```
Comment (hsliuustc0106) [inline tests/diffusion/models/magi2/test_native_compile_cuda.py:64]: The bitwise `torch.equal` assert here holds on the RTX 4070 (SM89) but not on Hopper: on 4x L20X (SM90, torch 2.13.0+cu130, triton 3.7.1 — the hardware class the MAGI-2 recipe qualifies), tiny-config bf16 video output differs in ~8–12.5% of elements with max abs diff 3e-8 (2–4 fp32 ulps; audio identical) — precisely the residual Triton `exp`/`sigmoid` ulp the PR body documents, amplified through routing. So the numerics behave as described, but this test will be red on any Hopper-class CI lane. Suggest an explicit tolerance guard on this comparison (e.g. `allclose` with atol≈1e-6, or asserting the max-ulp bound), keeping the bitwise assert where it provably holds.

Reproduced with `PYTHONPATH=. MAGI2_DETERMINISTIC=1 pytest tests/diffusion/models/magi2/test_native_compile_cuda.py -q` — 1 failed at the first `torch.equal(video, video_ref)`.

**Comment by hsliuustc0106** [comment]
Independent A/B on 4x L20X 140 GiB (SM90; vllm 0.28.0, torch 2.13.0+cu130, triton 3.7.1), resident SP4, 272p T2VA, 20 steps, seed 42, `MAGI2_DETERMINISTIC=1`, merge base `38abc0ba` vs head `9bde4cbf`, one warmup + 3 measured requests per arm (medians of the measured three):

| | base | head `--enforce-eager` | head compiled |
|---|---:|---:|---:|
| denoise loop (`sampler.sample`) | 19.4 s | 19.5 s | **14.6 s (−24.7%)** |
| client wall | 46.0 s | 46.2 s | 49.7 s |
| warmup request | 71.3 s | 69.3 s | 138.5 s |

- The A800 denoise-step win reproduces on Hopper: 0.97 → 0.73 s/step, −24.7% vs your −23.3%.
- The two eager-hygiene commits are byte-identical end-to-end: all 8 eager outputs across both arms share one sha256 (`a64858f4…`), and head-eager is perf-neutral (46.2 vs 46.0 s).
- Compiled numerics: PSNR 25.2 dB vs eager (inside the same-code eager spread), audio cosine 0.9959 (criterion ≥ 0.99), compiled run-to-run byte-identical (`74365add…` ×4).
- End-to-end wall did not improve on this host: ~20 s/request of per-request staging outside the profiled stages (conditioning-memory window + auxiliary VAE/audio staging, identical in every arm) plus one 6.8 s decode stall on the last compiled request dilute the denoise win. Your A800 table shows the e2e win (82 → 76.4 s), so e2e outcome is host-dependent while the denoise-step claim holds on both.
- First compiled request paid ~90 s of inductor compilation here (4 worker processes × 7 graphs), vs the 38 s reported on A800.

Same environment note as the inline thread: the tiny-config compiled-vs-eager comparison is not bitwise on SM90 (one-ulp Triton residual), so the CUDA test needs a tolerance guard for Hopper CI lanes.

File: vllm_omni/diffusion/models/magi2/pipeline_magi2.py (hunk 2 above)
Comment (cr-gao) [reply vllm_omni/diffusion/models/magi2/pipeline_magi2.py:698]: Added the log. `setup_compile()` now logs a warning when `diffusion_compile_granularity` is not `regional`, also added a corresponding test.

File: tests/diffusion/models/magi2/test_native_compile_cuda.py (hunk 3 above)
Comment (cr-gao) [reply tests/diffusion/models/magi2/test_native_compile_cuda.py]: Relaxed the compiled versus eager video comparison to `assert_close(atol=1e-6, rtol=0)`, the tolerance the audio comparison in the same test already used. 1e-6 should be a good number as the 3e-8 diff you measured would pass, while it still catches the numerical diff of 2.xe-4 it was written for.

**Comment by cr-gao** [comment]
> Independent A/B on 4x L20X 140 GiB (SM90; vllm 0.28.0, torch 2.13.0+cu130, triton 3.7.1), resident SP4, 272p T2VA, 20 steps, seed 42, `MAGI2_DETERMINISTIC=1`, merge base `38abc0ba` vs head `9bde4cbf`, one warmup + 3 measured requests per arm (medians of the measured three):
> 
> base	head `--enforce-eager`	head compiled
> denoise loop (`sampler.sample`)	19.4 s	19.5 s	**14.6 s (−24.7%)**
> client wall	46.0 s	46.2 s	49.7 s
> warmup request	71.3 s	69.3 s	138.5 s
> * The A800 denoise-step win reproduces on Hopper: 0.97 → 0.73 s/step, −24.7% vs your −23.3%.
> * The two eager-hygiene commits are byte-identical end-to-end: all 8 eager outputs across both arms share one sha256 (`a64858f4…`), and head-eager is perf-neutral (46.2 vs 46.0 s).
> * Compiled numerics: PSNR 25.2 dB vs eager (inside the same-code eager spread), audio cosine 0.9959 (criterion ≥ 0.99), compiled run-to-run byte-identical (`74365add…` ×4).
> * End-to-end wall did not improve on this host: ~20 s/request of per-request staging outside the profiled stages (conditioning-memory window + auxiliary VAE/audio staging, identical in every arm) plus one 6.8 s decode stall on the last compiled request dilute the denoise win. Your A800 table shows the e2e win (82 → 76.4 s), so e2e outcome is host-dependent while the denoise-step claim holds on both.
> * First compiled request paid ~90 s of inductor compilation here (4 worker processes × 7 graphs), vs the 38 s reported on A800.
> 
> Same environment note as the inline thread: the tiny-config compiled-vs-eager comparison is not bitwise on SM90 (one-ulp Triton residual), so the CUDA test needs a tolerance guard for Hopper CI lanes.

Got it. Both inline comments are addressed in bd8ac8711 and c5b895950.

**Review by hsliuustc0106 — APPROVED** [review]
lgtm
